### ISSUE TITLE
Please support @testing-library

### ISSUE THREAD

**Issue by dantman** [Codeception/CodeceptJS#3340]
#### What are you trying to achieve?

I want to write my tests using accessibility roles instead of hardcoding implementation details about how the html of my application is implemented. i.e. When I say in my test to "Click on the button with the label Login", it should not matter to the test whether the button is a `<button>Login</button>` or a `<div role="button" aria-label="Login">...</div>` and should behave the same either way.

This is a best practice supported by [`@testing-library`](https://testing-library.com/) which has libraries for nearly every test runner that CodeceptJS supports and also by Playwright using the [`role=` selector](https://playwright.dev/docs/selectors#role-selector) or using [`playwright-testing-library`](https://www.npmjs.com/package/playwright-testing-library).

#### What is stopping you from doing this now?

Even though almost every test runner has an associated testing-library for locating an element using whatever Selector or element object the test runner natively understands. Codecept's helpers do not accept these native elements and only accept Codecept's own Locator syntax. As a result it's not possible to use any `@testing-library` with Codecept JS' helpers.

##### A) Natively support `@testing-library`

New strict selectors like `role` could be added for each of `@testing-library`'s selectors (ByRole, ByLabelText, ByPlaceholderText, ByText, ByDisplayValue, ByAltText, ByTitle, ByTestId) and the `findElement` of each helpers module can use the related testing library as an optional peerDep if it is installed by the user.

##### B) Allow the user to specify a selector function that uses the helper's native API

Each helpers module could support an additional explicit selector with the test runner's own name that takes a user function that expects the user to return the test runner's own native Selector/element.

For example in TestCafe the following 
```js
I.click({ testcafe: () => new Selector('div')});
await within({ testcafe: () => new Selector('div')}, () => {
  I.click({ testcafe: (container) => container.find('button')});
});
```

This would allow any special locators or 3rd party library supporting a test runner to be used with Codecept. So the user could use this to use the related `@testing-library` library directly.

This is a little verbose of course. But that can easily be solved by writing custom functions to create the strict selector, e.g. `By.role('button', /Login/i)`.

**Comment by DavertMik** [comment]
CodeceptJS already implements most of features of testing-library so there is no reason to re-introduce them.
if you want to have role selector you can use customLocator plugin https://codecept.io/plugins/#customlocator

I'm closing this as this is not relevant to CodeceptJS

**Comment by dantman** [comment]
> CodeceptJS already implements most of features of testing-library so there is no reason to re-introduce them.

Codecept's selectors are almost nothing like testing-library's selectors. Codecept's selector code doesn't match browser behaviour for roles at all. The narrow selectors miss elements that the browser reports as exactly the same accessibility wise. While the wide ones consider completely irrelevant elements to be valid matches (fields match non-fields, clickable matches non-interactive elements). While also appearing to incorrectly implement aria-labelledby. They also seem to be case-sensitive when you don't always want case sensitive. And Codecept's selectors only handle vague concepts like "clickable", "field", "checkable", and "select". Instead of differentiating between the things the browser reports like buttons, dialogs, tabs, groups, menu items, etc.

> if you want to have role selector you can use customLocator plugin https://codecept.io/plugins/#customlocator

There is no way to implement a role selector like testing-library's with customLocator. From what I can see customLocator only lets you make css/xpath selectors. However the browser's accessible name/description matches used in things like `findByRole('button', { name: /Submit/i })` use a [complex algorithm](https://www.w3.org/TR/accname-1.2/) that cannot be implemented by a simple css/xpath selector. And roles are also not a simple attribute match, roles are sometimes explicit and sometimes implicit - both an explicitly specified role and an implicit roles are the exact same as far as the browser is concerned. e.g. If I have a test with a table cell role matching a native `<td>` element, it should still behave the exact same if I replace my data table library with one that happens to instead outputs a `<div role="cell">`.
